treerec/tskit/tree_sequence.c

```diff
@@ -0,0 +1,2438 @@
+    table_size_t offset, length;
+
+    if (index >= self->provenances.num_records) {
+        ret = MSP_ERR_OUT_OF_BOUNDS;
+        goto out;
+    }
+    provenance->id = (table_size_t) index;
+    offset = self->provenances.timestamp_offset[index];
+    length = self->provenances.timestamp_offset[index + 1] - offset;
+    provenance->timestamp = self->provenances.timestamp + offset;
+    provenance->timestamp_length = length;
+    offset = self->provenances.record_offset[index];
+    length = self->provenances.record_offset[index + 1] - offset;
+    provenance->record = self->provenances.record + offset;
+    provenance->record_length = length;
+out:
+    return ret;
+}
+
+int WARN_UNUSED
+tree_sequence_get_samples(tree_sequence_t *self, node_id_t **samples)
+{
+    *samples = self->samples;
+    return 0;
+}
+
+int WARN_UNUSED
+tree_sequence_get_sample_index_map(tree_sequence_t *self, node_id_t **sample_index_map)
+{
+    *sample_index_map = self->nodes.sample_index_map;
+    return 0;
+}
+
+int WARN_UNUSED
+tree_sequence_simplify(tree_sequence_t *self, node_id_t *samples, size_t num_samples,
+        int flags, tree_sequence_t *output, node_id_t *node_map)
+{
+    int ret = 0;
+    table_collection_t tables;
+
+    ret = tree_sequence_dump_tables(self, &tables, MSP_ALLOC_TABLES);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = table_collection_simplify(&tables, samples, num_samples, flags, node_map);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tree_sequence_load_tables(output, &tables, MSP_BUILD_INDEXES);
+out:
+    table_collection_free(&tables);
+    return ret;
+}
+
+// A little helper stuff for tree_sequence_compute_parents()...
+typedef struct  {
+	table_size_t site_index;		// the index in the site's mutation array for this mutation
+	node_id_t node_id;				// the node id for the node that this mutation was introduced in
+	mutation_id_t mutation_id;		// the mutation id for this mutation
+	
+	// branch group flags, used only for mutations that are in a branch group (i.e., one of a group along a single branch)
+	bool branch_group_checked;		// true if the mutation is found to be on the same branch as another mutation
+} node_map_t;
+
+static int node_map_compare_node_id(const void *p, const void *q)
+{
+	// sort by node_id, and by site_index below that so that we don't reorder entries within each branch group
+	int ln = ((node_map_t *)p)->node_id;
+	int rn = ((node_map_t *)q)->node_id;
+	
+	if (ln == rn)
+	{
+		int ls = ((node_map_t *)p)->site_index;
+		int rs = ((node_map_t *)q)->site_index;
+		
+		return (ls - rs);
+	}
+	
+	return (ln - rn);
+}
+
+static int node_map_compare_site_index(const void *p, const void *q)
+{
+	// sort by site_index (index in the site's mutation array), which fully determines the order
+	int l = ((node_map_t *)p)->site_index;
+	int r = ((node_map_t *)q)->site_index; 
+	return (l - r);
+}
+
+int WARN_UNUSED tree_sequence_compute_parents(tree_sequence_t *self, bool fix_missing_parents)
```

Comment:
should the second argument really be `bool`? i thought it was some `flags`? @bhaller

Reply:
The flags are for tree_sequence_load_tables().  In the present design, tree_sequence_compute_parents() takes a bool since that seems to be all it needs.  But all of this needs to be revisited in light of recent developments anyway...